File: problems/leetcode/0100/102/102.py
# 102. Binary Tree Level Order Traversal
# https://leetcode.com/problems/binary-tree-level-order-traversal/description/

from collections import deque
from typing import *


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        traversal = []

        q = deque()
        if root:
            q.append((root, 0))

        while len(q) > 0:
            node, level = q.popleft()

            while len(traversal) < level + 1:
                traversal.append([])

            traversal[level].append(node.val)

            if node.left:
                q.append((node.left, level + 1))
            if node.right:
                q.append((node.right, level + 1))

        return traversal

File: problems/leetcode/0100/102/102.cpp
// 102. Binary Tree Level Order Traversal
// https://leetcode.com/problems/binary-tree-level-order-traversal/

using namespace std;

#include <vector>

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
public:
  vector<vector<int>> levelOrderNodes;

  void appendToLevelOrderNodes(int val, int level) {
    if (levelOrderNodes.size() <= level) {
      while (levelOrderNodes.size() <= level) {
        vector<int> newLevel;
        levelOrderNodes.push_back(newLevel);
      }
    }
    levelOrderNodes[level].push_back(val);
  }

  void dfs(TreeNode *root, int level) {
    if (!root) {
      return;
    }
    appendToLevelOrderNodes(root->val, level);
    dfs(root->left, level + 1);
    dfs(root->right, level + 1);
  }

  vector<vector<int>> levelOrder(TreeNode *root) {
    dfs(root, 0);
    return levelOrderNodes;
  }
};
